src/Interpreters/Streaming/Aggregator.cpp

```diff
@@ -1566,59 +1445,60 @@ bool Aggregator::checkLimits(size_t result_size, bool & no_more_keys) const
-    AggregateColumnsData & aggregate_columns,
-    MutableColumns & final_aggregate_columns,
-    Arena * arena,
-    bool final,
-    bool clear_states) const
+Block Aggregator::convertToBlockImpl(
+    Method & method, Table & data, Arena * arena, Arenas & aggregates_pools, bool final, size_t rows, bool clear_states, AggregateStateType type) const
 {
     if (data.empty())
-        return;
-
-    if (key_columns.size() != params.keys_size)
-        throw Exception{"Aggregate. Unexpected key columns size.", ErrorCodes::LOGICAL_ERROR};
+    {
+        auto && out_cols = prepareOutputBlockColumns(params, aggregate_functions, getHeader(final), aggregates_pools, final, rows);
+        return {finalizeBlock(params, getHeader(final), std::move(out_cols), final, rows)};
+    }
 
-    std::vector<IColumn *> raw_key_columns;
-    raw_key_columns.reserve(key_columns.size());
-    for (auto & column : key_columns)
-        raw_key_columns.push_back(column.get());
+    Block res;
 
     if (final)
     {
 #if USE_EMBEDDED_COMPILER
-        if (compiled_aggregate_functions_holder)
+        /// TODO: So far not support compiled functions with expanded data
+        if (compiled_aggregate_functions_holder && !hasExpandedData())
         {
             static constexpr bool use_compiled_functions = !Method::low_cardinality_optimization;
-            convertToBlockImplFinal<Method, use_compiled_functions>(method, data, std::move(raw_key_columns), final_aggregate_columns, arena, clear_states);
+            assert(type == AggregateStateType::Normal);
+            res = convertToBlockImplFinal<Method, use_compiled_functions, AggregateStateType::Normal>(method, data, arena, aggregates_pools, rows, clear_states);
         }
         else
 #endif
         {
-            convertToBlockImplFinal<Method, false>(method, data, std::move(raw_key_columns), final_aggregate_columns, arena, clear_states);
+            if (type == AggregateStateType::OnlyUpdated)
+                res = convertToBlockImplFinal<Method, false, AggregateStateType::OnlyUpdated>(method, data, arena, aggregates_pools, rows, clear_states);
+            else if (type == AggregateStateType::OnlyRetracted)
+                res = convertToBlockImplFinal<Method, false, AggregateStateType::OnlyRetracted>(method, data, arena, aggregates_pools, rows, clear_states);
+            else
+                res = convertToBlockImplFinal<Method, false, AggregateStateType::Normal>(method, data, arena, aggregates_pools, rows, clear_states);
         }
     }
     else
     {
-        convertToBlockImplNotFinal(method, data, std::move(raw_key_columns), aggregate_columns);
+        assert(type == AggregateStateType::Normal);
+        res = convertToBlockImplNotFinal(method, data, aggregates_pools, rows);
     }
 
     /// In order to release memory early.
     /// proton: starts. For streaming aggr, we hold on to the states
-    if (clear_states)
+    if (clear_states && type == AggregateStateType::Normal)
```

Comment:
we may need revisit how we evaluate clear_states, instead of patching every where with additional different checking logic. It would be quite hard to maintain.

Reply:
This change should be removed, it was missed. 
`clear_states` already handles the branch in the outermost call, I will remove `&& type == AggregateStateType::Normal`